### PR TITLE
feat(restAPI): Added options request and verification function

### PR DESCRIPTION
**Pull request by Shruti3004**
## Description
#1938

`$this->options('', AuthController::class . ':optionsVerification');`
added the options request in /src/www/ui/api/index.php

added the optionsVerification request in /src/www/ui/api/Controllers/AuthController.php

`if(stristr($request->getMethod(), "options") !== false)`
setup the condition in /src/www/ui/api/Middlewares/RestAuthMiddleware.php

### CODE REVIEW

File: src/www/ui/api/index.php [hunk 1]
```diff
@@ -111,6 +111,7 @@ function (){
     $this->post('', UploadController::class . ':postUpload');
     $this->get('/{id:\\d+}/summary', UploadController::class . ':getUploadSummary');
     $this->get('/{id:\\d+}/licenses', UploadController::class . ':getUploadLicenses');
+    $this->options('/{params:.*}', AuthController::class . ':optionsVerification');
```
Comment (GMishx) [inline src/www/ui/api/index.php:114]: A much simpler way will to add something like following just before the middleware configuration:
```php
///////////////////////CORS OPTIONS////////////////
$app->options(VERSION_1 . '{routes:.+}', AuthController::class . ':optionsVerification');
```

File: src/www/ui/api/Controllers/AuthController.php [hunk 2]
```diff
@@ -55,7 +55,13 @@ public function getAuthHeaders($request, $response, $args)
     return $response->withHeader('Warning', $warningMessage)->withJson(
       $returnVal->getArray(), $returnVal->getCode());
   }
-
+  public function optionsVerification($request, $response, $args)
+  {
+    $Message = "Successful";
+    $returnVal = new Info(204, $Message, InfoType::INFO);
+    return $response->withJson(
+      $returnVal->getArray(), $returnVal->getCode());
```
Comment (GMishx) [inline src/www/ui/api/Controllers/AuthController.php:63]: [Response code 204](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/204) should not have a body. I would suggest following response:
```php
global $SysConf;
return $response->withStatus(204)
  ->withHeader('Access-Control-Allow-Origin', $SysConf['SYSCONFIG']['CorsOrigins'])
  ->withHeader('Access-Control-Allow-Headers', 'X-Requested-With, Content-Type, Accept, Origin, Authorization')
  ->withHeader('Access-Control-Allow-Methods', 'GET, POST, PUT, DELETE, PATCH, OPTIONS')
  ->withHeader('Access-Control-Allow-Credentials', 'true');
```

Please refer to [this cookbook](https://www.slimframework.com/docs/v3/cookbook/enable-cors.html) for more.

Comment (GMishx) [reply src/www/ui/api/Controllers/AuthController.php:63]: And to create the `CorsOrigins` option, please check another change at https://github.com/fossology/fossology/pull/1890/files#diff-24724248fe452296b11be6a8423331269089b3057511f90db0de4babb0b6eb33R395

**Review by GMishx — CHANGES_REQUESTED** [review]
Some suggestions.

File: src/www/ui/api/Controllers/AuthController.php (hunk 2 above)
Comment (Shruti3004) [reply src/www/ui/api/Controllers/AuthController.php]: @GMishx sir I have made the suggested changes but I am not getting this CorsOrigins option, I have set up this on my system with this method https://prezi.com/view/dKLarFhLWuQzfPOKpXkr/ please check this out I have documented here, and let me know the changes.

Comment (GMishx) [reply src/www/ui/api/Controllers/AuthController.php]: @Shruti3004 , did you had a look on the PR link I shared above?

File: src/www/ui/api/index.php [hunk 3]
```diff
@@ -110,37 +113,37 @@ function (){
     $this->put('/{id:\\d+}', UploadController::class . ':copyUpload');
     $this->post('', UploadController::class . ':postUpload');
     $this->get('/{id:\\d+}/summary', UploadController::class . ':getUploadSummary');
-    $this->get('/{id:\\d+}/licenses', UploadController::class . ':getUploadLicenses');
+    $this->get('/{id:\\d+}/licenses', UploadController::class . ':getUploadLicenses');
```
Comment (GMishx) [inline src/www/ui/api/index.php:116]: Please remove unnecessary whitespace changes here and in following lines.

Comment (Shruti3004) [reply src/www/ui/api/index.php:116]: > Please remove unnecessary whitespace changes here and in following lines.

Okay sir

File: src/www/ui/api/Controllers/AuthController.php (hunk 2 above)
Comment (Shruti3004) [reply src/www/ui/api/Controllers/AuthController.php]: Yes, Sir I have checked this pr but i am not able to refer it with cors

Comment (Shruti3004) [reply src/www/ui/api/Controllers/AuthController.php]: > @Shruti3004 , did you had a look on the PR link I shared above?

Yes sir, I have checked this pr but I am not able to refer it with cors.

Comment (GMishx) [reply src/www/ui/api/Controllers/AuthController.php]: I shared the link so you can understand how to create configuration variables in FOSSology. Using that variable, we can simply set the value for `Access-Control-Allow-Origin` header without changing anything in the source code.
You can access the configuration page on FOSSology at Admin > Customize menu.

Comment (Shruti3004) [reply src/www/ui/api/Controllers/AuthController.php]: > I shared the link so you can understand how to create configuration variables in FOSSology. Using that variable, we can simply set the value for `Access-Control-Allow-Origin` header without changing anything in the source code.
> You can access the configuration page on FOSSology at Admin > Customize menu.

Okay sir

File: src/www/ui/api/index.php (hunk 3 above)
Comment (GMishx) [reply src/www/ui/api/index.php:116]: Please, you do not need to call me sir :smiley:

Comment (Shruti3004) [reply src/www/ui/api/index.php:116]: > Please, you do not need to call me sir
Okay :sweat_smile:

File: src/www/ui/api/Controllers/AuthController.php (hunk 2 above)
Comment (Shruti3004) [reply src/www/ui/api/Controllers/AuthController.php]: @GMishx as far I get it, what we want to do is we are trying to make the CORS settings dynamic from the user interface so for that I need to add config variable, 
`$variable = "CorsOrigin";`
`$prompt = _('Access-Control-Allow-Origin');`
`$desc = _('[scheme]://[hostname]:[port]');`
`$valueArray[$variable] = array("'$variable'", "null", "'$prompt'",`
    `strval(CONFIG_TYPE_TEXT), "'CorsOrigin'", "1", "'$desc'", "null", "null");`

so sir let me know adding this variable would work or not and should I add localhost:3000 as default because we are going to do it on react

Comment (GMishx) [reply src/www/ui/api/Controllers/AuthController.php]: Sorry for making it confusing but yes, it does looks good. However, I would propose following changes:
```php
$variable = "CorsOrigins";
$prompt = _('Allowed origins for REST API');
$desc = _('[scheme]://[hostname]:[port], "*" for anywhere');
$valueArray[$variable] = array("'$variable'", "'localhost:3000'", "'$prompt'",
    strval(CONFIG_TYPE_TEXT), "'PAT'", "3", "'$desc'", "null", "null");
```

I just made some changes in the variable name and added the default value `localhost:3000`. Looks good to you?

Comment (Shruti3004) [reply src/www/ui/api/Controllers/AuthController.php]: > Sorry for making it confusing but yes, it does looks good. However, I would propose following changes:
> 
> ```
> $variable = "CorsOrigins";
> $prompt = _('Allowed origins for REST API');
> $desc = _('[scheme]://[hostname]:[port], "*" for anywhere');
> $valueArray[$variable] = array("'$variable'", "'localhost:3000'", "'$prompt'",
>     strval(CONFIG_TYPE_TEXT), "'PAT'", "3", "'$desc'", "null", "null");
> ```
> 
> I just made some changes in the variable name and added the default value `localhost:3000`. Looks good to you?

Yes, bhaiya looks good, let me make the following changes.

**Review by GMishx — APPROVED** [review]
Tested, working as expected.

**Comment by GMishx** [comment]
Please fix the error in code sniffer: 
image

Also, please squash the commits into one.
